sapp/ui/frontend/src/Filter.js

```diff
@@ -251,17 +290,50 @@ const SourceKinds = (
     <>
       <Label label="kind" />
       <Row gutter={gutter}>
-        <Col span={22}>
+        <Col span={6}>
           <Select
-            mode="multiple"
-            value={props.currentFilter.source_kinds}
-            options={allSourceKinds.map(value => ({value}))}
+            options={[{value: 'is'}, {value: 'matches'}]}
+            value={mode}
+            onChange={setMode}
             style={{width: '100%'}}
-            onChange={source_kinds =>
-              props.setCurrentFilter({...props.currentFilter, source_kinds})
-            }
           />
         </Col>
+        <Col span={16}>
+          {mode === 'is' ? (
+            <Select
+              mode="multiple"
+              value={props.currentFilter.source_kinds}
+              options={allSourceKinds.map(value => ({value}))}
+              style={{width: '100%'}}
+              onChange={source_kinds =>
+                props.setCurrentFilter({...props.currentFilter, source_kinds})
+              }
+            />
+          ): (
+            <Input
+              placeholder="regular expression"
+              style={{width: '100%'}}
+              value={inputValue}
+              onChange={event => {
+                const value = event.target.value;
+                setInputValue(value);
+                const source_kinds = allSourceKinds.filter(source_kind =>
+                  source_kind.match(value),
+                );
+                props.setCurrentFilter({...props.currentFilter, source_kinds});
+              }}
+              suffix={
+                <Tooltip
+                  title={(props.currentFilter.source_kinds || []).join('\n')}
+                  placement="bottom">
+                  <Text type="secondary" size="small">
+                    {(props.currentFilter.source_kinds || []).length}
+                  </Text>
+                </Tooltip>
+              }
+            />
+          )}
+        </Col>
```

Comment:
Looks like we repeat this pattern a couple of times. Do you think this could be factored out into its own component?

Reply:
I will try!